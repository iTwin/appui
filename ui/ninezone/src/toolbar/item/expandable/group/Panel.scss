/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/

@import "../../_variables";

.nz-toolbar-item-expandable-group-panel {
  $fill-color: rgba($background-5, 0.88);

  line-height: normal;
  background: $fill-color;
  padding-left: 16px;
  padding-right: 8px;
  padding-top: 8px;
  padding-bottom: 12px;
  border-radius: 4px;
}
