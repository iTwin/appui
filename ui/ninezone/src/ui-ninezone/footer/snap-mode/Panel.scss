/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
@import "@bentley/ui-core/lib/ui-core/style/themecolors";
@import "@bentley/ui-core/lib/ui-core/_z-index";
@import "../_variables";

.nz-footer-snapMode-panel {
  min-width: $nz-footer-popup-min-width;
  background-color: $buic-background-panel;
  border-radius: 3px;
  overflow: hidden;

  > .nz-title {
    line-height: 40px;
    font-size: 18px;
    text-align: center;
    color: $buic-text-color;
  }

  > .nz-snaps {
    display: grid;
    grid-gap: 15px;
    padding: 0 0 15px 5px;
  }
}
