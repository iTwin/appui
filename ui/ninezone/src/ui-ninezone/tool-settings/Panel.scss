/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
@import "~@bentley/ui-core/lib/ui-core/index";

.nz-toolSettings-panel {
  position: absolute;
  top: 100%;
  right: 0;
  background: $buic-background-1;
  border-color: $buic-background-5;
  border-width: 1px;
  border-style: solid;
  border-radius: 3px;
  padding: 0.5em;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}
