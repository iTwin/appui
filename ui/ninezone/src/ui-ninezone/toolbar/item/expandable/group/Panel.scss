/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/

@import "../../_variables";

.nz-toolbar-item-expandable-group-panel {
  $fill-color: $buic-background-panel;

  line-height: normal;
  background: $fill-color;
  padding-left: 16px;
  padding-right: 8px;
  padding-top: 8px;
  padding-bottom: 12px;
  border-radius: 4px;
  pointer-events: auto;
  border: 1px solid $buic-background-widget-stroke;
}
