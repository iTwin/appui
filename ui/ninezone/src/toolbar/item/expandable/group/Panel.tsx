/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Toolbar */

import * as classnames from "classnames";
import * as React from "react";
import CommonProps from "../../../../utilities/Props";
import "./Panel.scss";

/** Properties of [[Panel]] component. */
export interface PanelProps extends CommonProps {
  /** Panel content. */
  children?: React.ReactNode;
}

/** Basic panel. Used as base for [[Group]] and [[NestedGroup]] components. */
export default class Panel extends React.Component<PanelProps> {
  public render() {
    const className = classnames(
      "nz-toolbar-item-expandable-group-panel",
      this.props.className);

    return (
      <div
        className={className}
        style={this.props.style}
      >
        {this.props.children}
      </div>
    );
  }
}
