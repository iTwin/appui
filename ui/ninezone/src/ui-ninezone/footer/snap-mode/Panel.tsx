/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module SnapMode */

import * as classnames from "classnames";
import * as React from "react";
import { CommonProps } from "@bentley/ui-core";
import "./Panel.scss";

/** Properties of [[SnapModePanel]] component.
 * @beta
 */
export interface SnapModePanelProps extends CommonProps {
  /** Snap rows. I.e. [[Snap]] */
  children?: React.ReactNode;
  /** Dialog title. */
  title?: string;
}

/** Snap mode panel used with [[SnapMode]] component.
 * @note This is a presentational component and should be aligned with [[SnapMode]] component.
 * I.e. use [[FooterPopup]] to handle alignment.
 * @beta
 */
export class SnapModePanel extends React.PureComponent<SnapModePanelProps> {
  public render() {
    const className = classnames(
      "nz-footer-snapMode-panel",
      this.props.className);

    return (
      <div
        className={className}
        style={this.props.style}
      >
        <div className="nz-title">
          {this.props.title}
        </div>
        <div className="nz-snaps">
          {this.props.children}
        </div>
      </div>
    );
  }
}
