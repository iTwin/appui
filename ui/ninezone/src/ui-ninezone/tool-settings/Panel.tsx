/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module ToolSettings
 */

import "./Panel.scss";
import classnames from "classnames";
import * as React from "react";
import { ToolSettingProps } from "./Setting";

/** Properties of [[ToolSettingsOverflowPanel]] component.
 * @internal
 */
export interface ToolSettingsOverflowPanelProps extends ToolSettingProps {
  /** Panel content. */
  children?: React.ReactNode;
}

/** Displays overflown tool settings.
 * @internal
 */
// tslint:disable-next-line: variable-name
export const ToolSettingsOverflowPanel = React.forwardRef<HTMLDivElement, ToolSettingsOverflowPanelProps>((props, ref) => {
  const className = classnames(
    "nz-toolSettings-panel",
    props.className,
  );
  return (
    <div
      className={className}
      ref={ref}
      style={props.style}
    >
      {props.children}
    </div>
  );
});
