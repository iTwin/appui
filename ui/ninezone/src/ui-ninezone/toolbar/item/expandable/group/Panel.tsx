/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Toolbar */

import * as classnames from "classnames";
import * as React from "react";
import { CommonProps } from "../../../../utilities/Props";
import "./Panel.scss";

/** Properties of [[Panel]] component. */
export interface PanelProps extends CommonProps {
  /** Panel content. */
  children?: React.ReactNode;
}

/** Basic panel used in [[ExpandableItem]]. Used as base in [[Group]] and [[NestedGroup]] components. */
export class Panel extends React.PureComponent<PanelProps> {
  public render() {
    const className = classnames(
      "nz-toolbar-item-expandable-group-panel",
      this.props.className);

    return (
      <div
        className={className}
        style={this.props.style}
      >
        {this.props.children}
      </div>
    );
  }
}
