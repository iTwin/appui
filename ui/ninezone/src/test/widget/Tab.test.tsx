/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import * as React from "react";
import * as sinon from "sinon";
import { act, fireEvent, render } from "@testing-library/react";
import {
  addPanelWidget, addTab, createNineZoneState, NineZoneDispatch, NineZoneProvider, WIDGET_TAB_CLICK, WIDGET_TAB_DOUBLE_CLICK, WIDGET_TAB_DRAG_START,
  WidgetStateContext, WidgetTab,
} from "../../ui-ninezone";
import { PanelSideContext } from "../../ui-ninezone/widget-panels/Panel";
import { WidgetTabsEntryContext } from "../../ui-ninezone/widget/Tabs";
import { WidgetContext } from "../../ui-ninezone/widget/Widget";

describe("WidgetTab", () => {
  const sandbox = sinon.createSandbox();

  afterEach(() => {
    sandbox.restore();
  });

  it("should render active", () => {
    let nineZone = createNineZoneState();
    nineZone = addPanelWidget(nineZone, "left", "w1", { activeTabId: "t1" });
    nineZone = addTab(nineZone, "w1", "t1");
    const { container } = render(
      <NineZoneProvider
        state={nineZone}
        dispatch={sinon.spy()}
      >
        <WidgetStateContext.Provider value={nineZone.widgets.w1}>
          <WidgetTabsEntryContext.Provider value={{
            lastNotOverflown: false,
          }}>
            <WidgetTab tab={nineZone.tabs.t1} />
          </WidgetTabsEntryContext.Provider>
        </WidgetStateContext.Provider>
      </NineZoneProvider>,
    );
    container.firstChild!.should.matchSnapshot();
  });

  it("should render overflown", () => {
    let nineZone = createNineZoneState();
    nineZone = addPanelWidget(nineZone, "left", "w1");
    nineZone = addTab(nineZone, "w1", "t1");
    const { container } = render(
      <NineZoneProvider
        state={nineZone}
        dispatch={sinon.spy()}
      >
        <WidgetStateContext.Provider value={nineZone.widgets.w1}>
          <WidgetTabsEntryContext.Provider value={{
            lastNotOverflown: false,
          }}>
            <WidgetTab tab={nineZone.tabs.t1} />
          </WidgetTabsEntryContext.Provider>
        </WidgetStateContext.Provider>
      </NineZoneProvider>,
    );
    container.firstChild!.should.matchSnapshot();
  });

  it("should render minimized", () => {
    let nineZone = createNineZoneState();
    nineZone = addPanelWidget(nineZone, "left", "w1", { minimized: true });
    nineZone = addTab(nineZone, "w1", "t1");
    const { container } = render(
      <NineZoneProvider
        state={nineZone}
        dispatch={sinon.spy()}
      >
        <WidgetStateContext.Provider value={nineZone.widgets.w1}>
          <WidgetTabsEntryContext.Provider value={{
            lastNotOverflown: false,
          }}>
            <WidgetTab tab={nineZone.tabs.t1} />
          </WidgetTabsEntryContext.Provider>
        </WidgetStateContext.Provider>
      </NineZoneProvider>,
    );
    container.firstChild!.should.matchSnapshot();
  });

  it("should render first inactive", () => {
    let nineZone = createNineZoneState();
    nineZone = addPanelWidget(nineZone, "left", "w1", { activeTabId: "t1" });
    nineZone = addTab(nineZone, "w1", "t1");
    const { container } = render(
      <NineZoneProvider
        state={nineZone}
        dispatch={sinon.spy()}
      >
        <WidgetStateContext.Provider value={nineZone.widgets.w1}>
          <WidgetTabsEntryContext.Provider value={{
            lastNotOverflown: false,
          }}>
            <WidgetTab tab={nineZone.tabs.t1} firstInactive />
          </WidgetTabsEntryContext.Provider>
        </WidgetStateContext.Provider>
      </NineZoneProvider>,
    );
    container.firstChild!.should.matchSnapshot();
  });

  it("should render last not overflown", () => {
    let nineZone = createNineZoneState();
    nineZone = addPanelWidget(nineZone, "left", "w1", { activeTabId: "t1" });
    nineZone = addTab(nineZone, "w1", "t1");
    const { container } = render(
      <NineZoneProvider
        state={nineZone}
        dispatch={sinon.spy()}
      >
        <WidgetStateContext.Provider value={nineZone.widgets.w1}>
          <WidgetTabsEntryContext.Provider value={{
            lastNotOverflown: true,
          }}>
            <WidgetTab tab={nineZone.tabs.t1} />
          </WidgetTabsEntryContext.Provider>
        </WidgetStateContext.Provider>
      </NineZoneProvider>,
    );
    container.firstChild!.should.matchSnapshot();
  });

  it("should dispatch WIDGET_TAB_CLICK", () => {
    const fakeTimers = sandbox.useFakeTimers();
    const dispatch = sinon.stub<NineZoneDispatch>();
    let nineZone = createNineZoneState();
    nineZone = addPanelWidget(nineZone, "left", "w1");
    nineZone = addTab(nineZone, "w1", "t1");
    render(
      <NineZoneProvider
        state={nineZone}
        dispatch={dispatch}
      >
        <PanelSideContext.Provider value="left">
          <WidgetStateContext.Provider value={nineZone.widgets.w1}>
            <WidgetTabsEntryContext.Provider value={{
              lastNotOverflown: false,
            }}>
              <WidgetTab tab={nineZone.tabs.t1} />
            </WidgetTabsEntryContext.Provider>
          </WidgetStateContext.Provider>
        </PanelSideContext.Provider>
      </NineZoneProvider>,
    );
    const tab = document.getElementsByClassName("nz-widget-tab")[0];
    act(() => {
      fireEvent.pointerDown(tab);
      fireEvent.pointerUp(tab);
      fakeTimers.tick(300);
    });
    dispatch.calledOnceWithExactly(sinon.match({
      type: WIDGET_TAB_CLICK,
      side: "left",
      widgetId: "w1",
      id: "t1",
    })).should.true;
  });

  it("should dispatch WIDGET_TAB_DOUBLE_CLICK", () => {
    const fakeTimers = sandbox.useFakeTimers();
    const dispatch = sinon.stub<NineZoneDispatch>();
    let nineZone = createNineZoneState();
    nineZone = addPanelWidget(nineZone, "left", "w1");
    nineZone = addTab(nineZone, "w1", "t1");
    render(
      <NineZoneProvider
        state={nineZone}
        dispatch={dispatch}
      >
        <PanelSideContext.Provider value="left">
          <WidgetStateContext.Provider value={nineZone.widgets.w1}>
            <WidgetTabsEntryContext.Provider value={{
              lastNotOverflown: false,
            }}>
              <WidgetTab tab={nineZone.tabs.t1} />
            </WidgetTabsEntryContext.Provider>
          </WidgetStateContext.Provider>
        </PanelSideContext.Provider>
      </NineZoneProvider>,
    );
    const tab = document.getElementsByClassName("nz-widget-tab")[0];
    act(() => {
      fireEvent.pointerDown(tab);
      fireEvent.pointerUp(tab);
      fireEvent.pointerDown(tab);
      fireEvent.pointerUp(tab);
      fakeTimers.tick(300);
    });
    dispatch.calledOnceWithExactly(sinon.match({
      type: WIDGET_TAB_DOUBLE_CLICK,
      side: "left",
      widgetId: "w1",
      id: "t1",
    })).should.true;
  });

  it("should dispatch WIDGET_TAB_DRAG_START on pointer move", () => {
    const dispatch = sinon.stub<NineZoneDispatch>();
    let nineZone = createNineZoneState();
    nineZone = addPanelWidget(nineZone, "left", "w1");
    nineZone = addTab(nineZone, "w1", "t1");
    render(
      <NineZoneProvider
        state={nineZone}
        dispatch={dispatch}
      >
        <WidgetContext.Provider value={{ measure: () => ({ height: 0, width: 0 }) }}>
          <WidgetStateContext.Provider value={nineZone.widgets.w1}>
            <WidgetTab tab={nineZone.tabs.t1} />
          </WidgetStateContext.Provider>
        </WidgetContext.Provider>
      </NineZoneProvider>,
    );
    const tab = document.getElementsByClassName("nz-widget-tab")[0];
    act(() => {
      fireEvent.pointerDown(tab);

      const moveEvent = document.createEvent("MouseEvent");
      moveEvent.initEvent("pointermove");
      sinon.stub(moveEvent, "clientX").get(() => 10);
      sinon.stub(moveEvent, "clientY").get(() => 10);
      fireEvent(document, moveEvent);
    });
    dispatch.calledOnceWithExactly(sinon.match({
      type: WIDGET_TAB_DRAG_START,
      widgetId: "w1",
      id: "t1",
    })).should.true;
  });

  it("should not dispatch WIDGET_TAB_DRAG_START on pointer move if pointer moved less than 10px", () => {
    const dispatch = sinon.stub<NineZoneDispatch>();
    let nineZone = createNineZoneState();
    nineZone = addPanelWidget(nineZone, "left", "w1");
    nineZone = addTab(nineZone, "w1", "t1");
    render(
      <NineZoneProvider
        state={nineZone}
        dispatch={dispatch}
      >
        <WidgetStateContext.Provider value={nineZone.widgets.w1}>
          <WidgetTab tab={nineZone.tabs.t1} />
        </WidgetStateContext.Provider>
      </NineZoneProvider>,
    );
    const tab = document.getElementsByClassName("nz-widget-tab")[0];
    act(() => {
      fireEvent.pointerDown(tab);

      const moveEvent = document.createEvent("MouseEvent");
      moveEvent.initEvent("pointermove");
      sinon.stub(moveEvent, "clientX").get(() => 5);
      sinon.stub(moveEvent, "clientY").get(() => 0);
      fireEvent(document, moveEvent);
    });
    dispatch.notCalled.should.true;
  });
});
