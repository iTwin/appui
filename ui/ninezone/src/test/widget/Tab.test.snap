// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetTab should render active 1`] = `
<div
  class="nz-widget-tab nz-active"
>
  <span
    title=""
  />
</div>
`;

exports[`WidgetTab should render first inactive 1`] = `
<div
  class="nz-widget-tab nz-active nz-first-inactive"
>
  <span
    title=""
  />
</div>
`;

exports[`WidgetTab should render last not overflown 1`] = `
<div
  class="nz-widget-tab nz-active nz-last-not-overflown"
>
  <span
    title=""
  />
</div>
`;

exports[`WidgetTab should render minimized 1`] = `
<div
  class="nz-widget-tab nz-minimized"
>
  <span
    title=""
  />
</div>
`;

exports[`WidgetTab should render overflown 1`] = `
<div
  class="nz-widget-tab"
>
  <span
    title=""
  />
</div>
`;
